llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -2671,20 +2837,22 @@ bool ModuleAddressSanitizer::instrumentModule(Module &M) {
   // Put the constructor and destructor in comdat if both
   // (1) global instrumentation is not TU-specific
   // (2) target is ELF.
-  if (UseCtorComdat && TargetTriple.isOSBinFormatELF() && CtorComdat) {
-    if (AsanCtorFunction) {
-      AsanCtorFunction->setComdat(M.getOrInsertComdat(kAsanModuleCtorName));
-      appendToGlobalCtors(M, AsanCtorFunction, Priority, AsanCtorFunction);
-    }
-    if (AsanDtorFunction) {
-      AsanDtorFunction->setComdat(M.getOrInsertComdat(kAsanModuleDtorName));
-      appendToGlobalDtors(M, AsanDtorFunction, Priority, AsanDtorFunction);
+  if (!TargetTriple.isSPIR()) { // SPIR kernel needn't AsanCtorFunction & AsanDtorFunction
```

Comment:
This should be converted to an early exit instead:

```c++
if (TargetTriple.isSPIR()) // your comment here
  return;
// rest of the code goes as-is, without any modifications
```

Maybe even better, we can modify `CtorComdat`, so our customization would be less fragile in case of future upstream changes:
```c++
if (TargetTriple.isSPIR()) // your comment here
  CtorComdat = true;
```

Reply:
Thanks for your tips, which inspired me to found another way to make it more stable.